src/scene/batching.js

```diff
@@ -0,0 +1,837 @@
+     */
+    var Batch = function (meshInstances, dynamic) {
+        this.origMeshInstances = meshInstances;
+        this._aabb = new pc.BoundingBox();
+        this.meshInstance = null;
+        this.model = null;
+        this.dynamic = dynamic;
+    };
+
+    var BatchGroup = function (id, name, dynamic, maxAabbSize) {
+        this.dynamic = dynamic;
+        this.maxAabbSize = maxAabbSize;
+        this.id = id;
+        this.name = name;
+    };
+
+    // Modified SkinInstance for batching
+    // Doesn't contain bind matrices, simplier
+    var SkinBatchInstance = function (device, nodes, rootNode) {
+        this.device = device;
+        this.rootNode = rootNode;
+        this._dirty = true;
+
+        // Unique per clone
+        this.bones = nodes;
+
+        var numBones = nodes.length;
+
+        if (device.supportsBoneTextures) {
+            var size;
+            if (numBones > 256)
+                size = 64;
+            else if (numBones > 64)
+                size = 32;
+            else if (numBones > 16)
+                size = 16;
+            else
+                size = 8;
+
+            this.boneTexture = new pc.Texture(device, {
+                width: size,
+                height: size,
+                format: pc.PIXELFORMAT_RGBA32F,
+                mipmaps: false,
+                minFilter: pc.FILTER_NEAREST,
+                magFilter: pc.FILTER_NEAREST
+            });
+            this.matrixPalette = this.boneTexture.lock();
+        } else {
+            this.matrixPalette = new Float32Array(numBones * 16);
+        }
+    };
+
+    SkinBatchInstance.prototype = {
+
+        updateMatrices: function () {
+        },
+
+        updateMatrixPalette: function () {
+            var pe;
+            var mp = this.matrixPalette;
+            var base;
+
+            for (var i = this.bones.length - 1; i >= 0; i--) {
+                pe = this.bones[i].getWorldTransform().data;
+
+                // Copy the matrix into the palette, ready to be sent to the vertex shader
+                base = i * 16;
+                mp[base] = pe[0];
+                mp[base + 1] = pe[1];
+                mp[base + 2] = pe[2];
+                mp[base + 3] = pe[3];
+                mp[base + 4] = pe[4];
+                mp[base + 5] = pe[5];
+                mp[base + 6] = pe[6];
+                mp[base + 7] = pe[7];
+                mp[base + 8] = pe[8];
+                mp[base + 9] = pe[9];
+                mp[base + 10] = pe[10];
+                mp[base + 11] = pe[11];
+                mp[base + 12] = pe[12];
+                mp[base + 13] = pe[13];
+                mp[base + 14] = pe[14];
+                mp[base + 15] = pe[15];
+            }
+
+            if (this.device.supportsBoneTextures) {
+                this.boneTexture.lock();
+                this.boneTexture.unlock();
+            }
+        }
+    };
+
+    /**
+     * @name pc.BatchManager
```

Comment:
Does this need to be private? At least for now?

Reply:
Why?